### PR TITLE
[GCS FT] Redis e2e cleanup check

### PR DESCRIPTION
**Pull request by rueian**
## Why are these changes needed?

Follow-ups for https://github.com/ray-project/kuberay/issues/2696#issuecomment-2599448835 and resolves https://github.com/ray-project/kuberay/issues/2764.

* Add e2e tests to check if Redis is empty at the end of each GCS FT test.
* Fix Redis cleanup job to accept empty passwords.

## Checks

- [x] I've made sure the tests are passing.
- Testing Strategy
   - [x] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: ray-operator/test/e2e/support.go [hunk 1]
```diff
@@ -177,26 +180,20 @@ func jobSubmitterPodTemplateApplyConfiguration() *corev1ac.PodTemplateSpecApplyC
...
 	if password != "" {
 		redisContainer.WithCommand("redis-server", "--requirepass", password)
+		dbSizeCmd = []string{"redis-cli", "--no-auth-warning", "-a", password, "DBSIZE"}
 	}
 
-	_, err := t.Client().Core().AppsV1().Deployments(namespace).Apply(
+	_, err := t.Client().Core().CoreV1().Pods(namespace).Apply(
 		t.Ctx(),
-		appsv1ac.Deployment("redis", namespace).
-			WithSpec(appsv1ac.DeploymentSpec().
```
Comment (rueian) [inline ray-operator/test/e2e/support.go:190]: no need for us to use Deployment for Redis in tests. just use Pod instead.

File: ray-operator/test/e2e/support.go [hunk 2]
```diff
@@ -213,4 +210,46 @@ func deployRedis(t Test, namespace string, password string) {
...
+
+		return strings.TrimSpace(stdout.String() + stderr.String())
+	}
+	return func() {
+		var output string
+		for i := 0; i < 30; i++ {
+			if output = checkDBSize(); output == "0" {
+				return
+			}
+			time.Sleep(time.Second)
```
Comment (rueian) [inline ray-operator/test/e2e/support.go:251]: Wait for the cleanup job.

File: ray-operator/test/e2e/support.go [hunk 3]
```diff
@@ -213,4 +210,46 @@ func deployRedis(t Test, namespace string, password string) {
 		TestApplyOptions,
 	)
 	assert.NoError(t.T(), err)
+
+	checkDBSize := func() string {
```
Comment (kevin85421) [inline ray-operator/test/e2e/support.go:214]: Use `ExecPodCmd` instead?

File: ray-operator/test/e2e/support.go [hunk 4]
```diff
@@ -213,4 +210,46 @@ func deployRedis(t Test, namespace string, password string) {
...
+			Stdout: stdout,
+			Stderr: stderr,
+		})
+		if err != nil {
+			t.T().Fatalf("failed to execute: %v", err)
+		}
+
+		return strings.TrimSpace(stdout.String() + stderr.String())
+	}
+	return func() {
```
Comment (kevin85421) [inline ray-operator/test/e2e/support.go:245]: Maybe only return `checkDBSize` and use `Eventually` in `raycluster_gcsft_test.go`?

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 5]
```diff
@@ -1230,7 +1230,7 @@ func (r *RayClusterReconciler) buildRedisCleanupJob(ctx context.Context, instanc
 			"redis_address = os.getenv('RAY_REDIS_ADDRESS', '').split(',')[0]; " +
 			"redis_address = redis_address if '://' in redis_address else 'redis://' + redis_address; " +
 			"parsed = urlparse(redis_address); " +
-			"sys.exit(1) if not cleanup_redis_storage(host=parsed.hostname, port=parsed.port, password=os.getenv('REDIS_PASSWORD', parsed.password), use_ssl=parsed.scheme=='rediss', storage_namespace=os.getenv('RAY_external_storage_namespace')) else None\"",
+			"sys.exit(1) if not cleanup_redis_storage(host=parsed.hostname, port=parsed.port, password=os.getenv('REDIS_PASSWORD', parsed.password or ''), use_ssl=parsed.scheme=='rediss', storage_namespace=os.getenv('RAY_external_storage_namespace')) else None\"",
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:1233]: would you mind explaining the change?

File: ray-operator/test/e2e/raycluster_gcsft_test.go [hunk 6]
```diff
@@ -81,7 +81,7 @@ func TestGcsFaultToleranceOptions(t *testing.T) {
 			g := NewWithT(t)
 			namespace := test.NewTestNamespace()
 
-			deployRedis(test, namespace.Name, tc.redisPassword)
+			defer deployRedis(test, namespace.Name, tc.redisPassword)()
```
Comment (kevin85421) [inline ray-operator/test/e2e/raycluster_gcsft_test.go:84]: what will the stack trace look like if the check in `defer` fails? If it is not easy to read, maybe we can return the clean up function and explicitly call it with `Eventually` at the end of the test logic?

Comment (rueian) [reply ray-operator/test/e2e/raycluster_gcsft_test.go]: This trace is the defer approach:

```sh
=== RUN   TestGcsFaultToleranceAnnotations/GCS_FT_without_redis_password
Modified Ray Image to: rayproject/ray:2.40.0-aarch64 for ARM chips
    raycluster_gcsft_test.go:211: Created RayCluster test-ns-72s65/raycluster-gcsft successfully
    raycluster_gcsft_test.go:213: Waiting for RayCluster test-ns-72s65/raycluster-gcsft to become ready
    raycluster_gcsft_test.go:217: Verifying environment variables on Head Pod
    support.go:253: failed cleanup redis expect 0 but got: 3
    test.go:110: Retrieving Pod Container test-ns-72s65/redis/redis logs
    test.go:98: Creating ephemeral output directory as KUBERAY_TEST_OUTPUT_DIR env variable is unset
    test.go:101: Output directory has been created at: /var/folders/sw/cyfhnvns2hv82r3fj1sgwsb00000gn/T/TestGcsFaultToleranceAnnotationsGCS_FT_without_redis_password2730163886/001
--- FAIL: TestGcsFaultToleranceAnnotations/GCS_FT_without_redis_password (48.50s)
```

This one is the trace for `ExecPodCmd` + `Eventually`.
```sh
=== RUN   TestGcsFaultToleranceAnnotations/GCS_FT_without_redis_password
Modified Ray Image to: rayproject/ray:2.40.0-aarch64 for ARM chips
    raycluster_gcsft_test.go:214: Created RayCluster test-ns-s257l/raycluster-gcsft successfully
    raycluster_gcsft_test.go:216: Waiting for RayCluster test-ns-s257l/raycluster-gcsft to become ready
    raycluster_gcsft_test.go:220: Verifying environment variables on Head Pod
    core.go:88: Executing command: [redis-cli --no-auth-warning DBSIZE]
    core.go:101: Command stdout: 3
    core.go:102: Command stderr: 
    ...
    core.go:88: Executing command: [redis-cli --no-auth-warning DBSIZE]
    core.go:101: Command stdout: 3
    core.go:102: Command stderr: 
    core.go:88: Executing command: [redis-cli --no-auth-warning DBSIZE]
    core.go:101: Command stdout: 3
    core.go:102: Command stderr: 
    raycluster_gcsft_test.go:236: 
        Timed out after 30.072s.
        Expected
            <string>: 3
        to be equivalent to
            <string>: 0
    test.go:110: Retrieving Pod Container test-ns-s257l/redis/redis logs
    test.go:98: Creating ephemeral output directory as KUBERAY_TEST_OUTPUT_DIR env variable is unset
    test.go:101: Output directory has been created at: /var/folders/sw/cyfhnvns2hv82r3fj1sgwsb00000gn/T/TestGcsFaultToleranceAnnotationsGCS_FT_without_redis_password315262234/001
--- FAIL: TestGcsFaultToleranceAnnotations/GCS_FT_without_redis_password (51.40s)
```

File: ray-operator/test/e2e/support.go (hunk 3 above)
Comment (rueian) [reply ray-operator/test/e2e/support.go]: 👍

File: ray-operator/test/e2e/support.go (hunk 4 above)
Comment (rueian) [reply ray-operator/test/e2e/support.go]: 👍

**Review by kevin85421 — APPROVED** [review]
